### ISSUE TITLE
Wrong workspace folder selection when cwd and ancestor directories also match rootPatterns

### ISSUE THREAD

**Issue by alansartorio** [neoclide/coc.nvim#4554]
## Result from CocInfo

```
## versions

vim version: NVIM v0.8.3
node version: v19.7.0
coc.nvim version: 0.0.82-b28b8dc 2023-03-02 11:47:57 +0800
coc.nvim directory: /home/alan/.local/share/nvim/plugged/coc.nvim
term: alacritty
platform: linux

## Log of coc.nvim

2023-03-07T00:04:49.579 INFO (pid:29611) [configurations] - Add folder configuration from cwd: /home/alan/.local/share/nvim/plugged/coc.nvim/.vim/coc-settings.json
2023-03-07T00:04:49.732 INFO (pid:29611) [plugin] - coc.nvim initialized with node: v19.7.0 after 203
2023-03-07T00:06:07.900 INFO (pid:29611) [services] - service tsserver started
2023-03-07T00:09:33.379 INFO (pid:29611) [attach] - receive notification: jumpReferences []
2023-03-07T00:09:33.418 INFO (pid:29611) [attach] - receive notification: openList [ '--normal', '--auto-preview', 'location' ]
2023-03-07T00:10:43.599 INFO (pid:29611) [snippets-session] - Content change after snippet, cancel snippet session
2023-03-07T00:11:19.625 INFO (pid:29611) [attach] - Request action: doKeymap [ 'i-PGJzPg==' ]
2023-03-07T00:11:20.474 INFO (pid:29611) [attach] - Request action: doKeymap [ 'i-PGJzPg==' ]
2023-03-07T00:11:20.636 INFO (pid:29611) [attach] - Request action: doKeymap [ 'i-PGJzPg==' ]
2023-03-07T00:11:20.783 INFO (pid:29611) [attach] - Request action: doKeymap [ 'i-PGJzPg==' ]
2023-03-07T00:12:45.764 INFO (pid:29611) [snippets-session] - Content change before snippet, reset snippet position
2023-03-07T00:12:49.354 INFO (pid:29611) [snippets-session] - Unable to find changed placeholder, cancel snippet session
2023-03-07T00:12:53.036 INFO (pid:29611) [attach] - Request action: doKeymap [ 'i-PGJzPg==' ]
2023-03-07T00:12:57.816 INFO (pid:29611) [attach] - Request action: doKeymap [ 'i-ew==' ]
2023-03-07T00:13:16.987 INFO (pid:29611) [attach] - Request action: doKeymap [ 'i-ew==' ]
2023-03-07T00:13:26.301 INFO (pid:29611) [attach] - receive notification: startCompletion [ {} ]
2023-03-07T00:13:37.388 INFO (pid:29611) [snippets-session] - Jump to final placeholder, cancelling snippet session

```

## Describe the bug

When opening a directory with nvim, even if it contains a file that matches with rootPatterns, it doesn't get chosen as workspace folder if there's any ancestor directory that also matches.

## Reproduce the bug

Having the following folder structure
```
test-coc
├── .git
└── inner
    └── .git
```

1. stand in `test-coc/inner` 
2. do `nvim .`
3. run `:echo WorkspaceFolders`

It returns `test-coc` instead of `inner`. I think this is not intended.

## Possible solution

In case the current document is a directory, not getting the dirname here: https://github.com/neoclide/coc.nvim/blob/955a3754114f5f1c0cf771234b0c1bd791314c52/src/core/workspaceFolder.ts#L160 would fix it for the example I presented.

```diff
diff --git a/src/core/workspaceFolder.ts b/src/core/workspaceFolder.ts
index d93c57b..87200dc 100644
--- a/src/core/workspaceFolder.ts
+++ b/src/core/workspaceFolder.ts
@@ -10,7 +10,7 @@ import { getConditionValue } from '../util'
 import { distinct, isFalsyOrEmpty, toArray } from '../util/array'
 import { isCancellationError } from '../util/errors'
 import { Extensions as ExtensionsInfo, IExtensionRegistry } from '../util/extensionRegistry'
-import { checkFolder, isFolderIgnored, isParentFolder, resolveRoot } from '../util/fs'
+import { checkFolder, isDirectory, isFolderIgnored, isParentFolder, resolveRoot } from '../util/fs'
 import { path } from '../util/node'
 import { toObject } from '../util/object'
 import { CancellationToken, CancellationTokenSource, Emitter, Event } from '../util/protocol'
@@ -157,7 +157,12 @@ export default class WorkspaceFolderController {
     let u = URI.parse(document.uri)
     let curr = this.getWorkspaceFolder(u)
     if (curr) return URI.parse(curr.uri).fsPath
-    let dir = path.dirname(u.fsPath)
+    let dir: string;
+    if (isDirectory(u.fsPath)) {
+      dir = path.normalize(u.fsPath)
+    } else {
+      dir = path.dirname(u.fsPath)
+    }
     let { ignoredFiletypes, ignoredFolders, workspaceFolderCheckCwd, workspaceFolderFallbackCwd, bottomUpFiletypes } = this.config
     if (ignoredFiletypes?.includes(document.filetype)) return null
     ignoredFolders = Array.isArray(ignoredFolders) ? ignoredFolders.filter(s => s && s.length > 0).map(s => expand(s)) : []
```

**Comment by fannheyward** [comment]
> `nvim .`

You're editing a file named `.`, not current directory.

**Comment by alansartorio** [comment]
But `.` refers to current directory, `nvim .` opens netrw in current directory.

**Comment by fannheyward** [comment]
> nvim . opens netrw in current directory

Something is broken in my test, will check it.

**Comment by fannheyward** [comment]
> nvim . opens netrw in current directory

In my tests, `nvim .` didn't open netrw, but I think we should check directory is needed.

**Comment by untitaker** [comment]
I am also running into this issue now. I am unable to use rust-analyzer with ra-multiplex and coc-rust-analyzer in certain repositories that mix both python and rust code and do not share the same workspace root.

For example, check out https://github.com/getsentry/snuba. It has a workspace structure like this:

```
./setup.py
./rust_snuba/
./rust_snuba/Cargo.toml
```

Even when I am cd'd into `rust_snuba/`, the effective workspace (LSP client cwd) is sent as the git root `./`, even with a configuration like:

```
  "workspace.rootPatterns": ["Cargo.toml", ".git", ".envrc", "setup.py", "pyproject.toml"],
```

Then rust-analyzer is spawned in the git root, where it cannot find a Rust project at all and all static analysis fails.

**Comment by untitaker** [comment]
ignore me, i am not even editing the current directory. so the fix at #4566 does not even apply to me
